### PR TITLE
Support for ILM 

### PR DESCRIPTION
**Pull request by dvaldivia**

### CODE REVIEW

File: restapi/admin_tiers.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+	"github.com/minio/console/restapi/operations"
+	"github.com/minio/console/restapi/operations/admin_api"
+)
+
+func registerAdminTiersHandlers(api *operations.ConsoleAPI) {
+	// return a list of notification endpoints
+	api.AdminAPITiersListHandler = admin_api.TiersListHandlerFunc(func(params admin_api.TiersListParams, session *models.Principal) middleware.Responder {
+		tierList, err := getTiersResponse(session)
+		if err != nil {
+			return admin_api.NewNotificationEndpointListDefault(int(err.Code)).WithPayload(err)
```
Comment (cesnietor) [inline restapi/admin_tiers.go:37]: wrong Response type

File: restapi/admin_tiers.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+		err := getEditTierCredentialsResponse(session, &params)
+		if err != nil {
+			return admin_api.NewEditTierCredentialsDefault(int(err.Code)).WithPayload(err)
+		}
+		return admin_api.NewEditTierCredentialsOK()
+	})
+
+}
+
+// getNotificationEndpoints invokes admin info and returns a list of notification endpoints
```
Comment (cesnietor) [inline restapi/admin_tiers.go:68]: wrong comments

File: restapi/admin_tiers.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+		}
+	}
+
+	// build response
+	return &models.TierListResponse{
+		Items: tiersList,
+	}, nil
+}
+
+// getNotificationEndpointsResponse returns a list of notification endpoints in the instance
```
Comment (cesnietor) [inline restapi/admin_tiers.go:133]: wrong comment

File: restapi/admin_tiers.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+			Type: madmin.Unsupported,
+		}
+
+	}
+
+	err = client.addTier(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return params.Body, nil
```
Comment (cesnietor) [inline restapi/admin_tiers.go:210]: we shouldn't return the same input, we can probably skip returning it and just return error.

File: restapi/admin_tiers.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+// getNotificationEndpointsResponse returns a list of notification endpoints in the instance
```
Comment (cesnietor) [inline restapi/admin_tiers.go:213]: wrong comments

File: restapi/admin_tiers.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	// create a minioClient interface implementation
+	// defining the client to be used
+	adminClient := adminClient{client: mAdmin}
+	// 20 seconds timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	// serialize output
+	addTierResp, err := getTier(ctx, adminClient, params)
+	if err != nil {
+		log.Println("error here")
```
Comment (cesnietor) [inline restapi/admin_tiers.go:314]:
```suggestion
```

File: restapi/admin_tiers.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+	// create a minioClient interface implementation
+	// defining the client to be used
+	adminClient := adminClient{client: mAdmin}
+	// 20 seconds timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	// serialize output
+	addTierResp, err := addTier(ctx, adminClient, params)
+	if err != nil {
+		log.Println("error here")
```
Comment (cesnietor) [inline restapi/admin_tiers.go:228]:
```suggestion
```

File: restapi/admin_tiers.go [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+func registerAdminTiersHandlers(api *operations.ConsoleAPI) {
+	// return a list of notification endpoints
+	api.AdminAPITiersListHandler = admin_api.TiersListHandlerFunc(func(params admin_api.TiersListParams, session *models.Principal) middleware.Responder {
+		tierList, err := getTiersResponse(session)
+		if err != nil {
+			return admin_api.NewNotificationEndpointListDefault(int(err.Code)).WithPayload(err)
+		}
+		return admin_api.NewTiersListOK().WithPayload(tierList)
+	})
+	// add a new notification endpoints
```
Comment (cesnietor) [inline restapi/admin_tiers.go:41]: wrong comment

File: restapi/admin_tiers.go [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+	api.AdminAPIAddTierHandler = admin_api.AddTierHandlerFunc(func(params admin_api.AddTierParams, session *models.Principal) middleware.Responder {
+		notifEndpoints, err := getAddTierResponse(session, &params)
+		if err != nil {
+			return admin_api.NewAddTierDefault(int(err.Code)).WithPayload(err)
+		}
+		return admin_api.NewAddTierCreated().WithPayload(notifEndpoints)
+	})
+	// add a new notification endpoints
```
Comment (cesnietor) [inline restapi/admin_tiers.go:49]: wrong comment

File: restapi/admin_tiers.go [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+				},
+			}, nil
+		}
+	}
+
+	// build response
+	return nil, ErrorGenericNotFound
+}
+
+// getNotificationEndpointsResponse returns a list of notification endpoints in the instance
```
Comment (cesnietor) [inline restapi/admin_tiers.go:299]: wrong comment

File: restapi/admin_tiers.go [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+		CredsJSON: []byte(params.Body.Creds),
+	}
+	err := client.editTierCreds(ctx, params.Name, creds)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getNotificationEndpointsResponse returns a list of notification endpoints in the instance
```
Comment (cesnietor) [inline restapi/admin_tiers.go:333]: wrong comments

File: restapi/admin_tiers.go [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+	creds := madmin.TierCreds{
+		AccessKey: params.Body.AccessKey,
+		SecretKey: params.Body.SecretKey,
+		CredsJSON: []byte(params.Body.Creds),
+	}
+	err := client.editTierCreds(ctx, params.Name, creds)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (cesnietor) [inline restapi/admin_tiers.go:330]:
```suggestion
	return client.editTierCreds(ctx, params.Name, creds)
```

File: restapi/admin_tiers.go [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+	// create a minioClient interface implementation
+	// defining the client to be used
+	adminClient := adminClient{client: mAdmin}
+	// 20 seconds timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	// serialize output
+	err = editTierCredentials(ctx, adminClient, params)
+	if err != nil {
+		log.Println("error here")
```
Comment (cesnietor) [inline restapi/admin_tiers.go:348]:
```suggestion
```

File: restapi/client-admin.go [hunk 14]
```diff
@@ -298,6 +302,21 @@ func (ac adminClient) serverHealthInfo(ctx context.Context, healthDataTypes []ma
 	return ac.client.ServerHealthInfo(ctx, healthDataTypes, deadline)
 }
 
+// implements madmin.listTiers()
+func (ac adminClient) listTiers(ctx context.Context) ([]*madmin.TierConfig, error) {
+	return ac.client.ListTiers(ctx)
+}
+
+// implements madmin.ServerInfo()
```
Comment (cesnietor) [inline restapi/client-admin.go:310]: wrong comment

File: restapi/client-admin.go [hunk 15]
```diff
@@ -298,6 +302,21 @@ func (ac adminClient) serverHealthInfo(ctx context.Context, healthDataTypes []ma
...
+func (ac adminClient) addTier(ctx context.Context, cfg *madmin.TierConfig) error {
+	return ac.client.AddTier(ctx, cfg)
+}
+
+// implements madmin.ServerInfo()
```
Comment (cesnietor) [inline restapi/client-admin.go:315]: wrong comment

File: restapi/user_buckets_lifecycle.go [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+			return user_api.NewGetBucketLifecycleDefault(int(err.Code)).WithPayload(err)
+		}
+		return user_api.NewGetBucketLifecycleOK().WithPayload(listBucketLifecycleResponse)
+	})
+	api.UserAPIAddBucketLifecycleHandler = user_api.AddBucketLifecycleHandlerFunc(func(params user_api.AddBucketLifecycleParams, session *models.Principal) middleware.Responder {
+		err := getAddBucketLifecycleResponse(session, params)
+		if err != nil {
+			return user_api.NewAddBucketLifecycleDefault(int(err.Code)).WithPayload(err)
+		}
+		return user_api.NewAddBucketReplicationCreated()
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:53]: wrong response?

File: restapi/user_buckets_lifecycle.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+			Expiration: &models.ExpirationResponse{Date: rule.Expiration.Date.Format(time.RFC3339), Days: int64(rule.Expiration.Days), DeleteMarker: rule.Expiration.DeleteMarker.IsEnabled()},
+			Transition: &models.TransitionResponse{Date: rule.Transition.Date.Format(time.RFC3339), Days: int64(rule.Transition.Days), StorageClass: rule.Transition.StorageClass},
+			Tags:       tags,
+		})
+	}
+
+	return rules, nil
+}
+
+// getListBucketsResponse performs getBucketLifecycle() and serializes it to the handler's output
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:90]: wrong comment

File: restapi/user_buckets_lifecycle.go [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+	minioClient := minioClient{client: mClient}
+
+	bucketEvents, err := getBucketLifecycle(ctx, minioClient, params.BucketName)
+	if err != nil {
+		return nil, prepareError(err)
+	}
+	// serialize output
+	lifecycleBucketsResponse := &models.BucketLifecycleResponse{
+		Lifecycle: bucketEvents,
+	}
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:108]: please make this inside `getBucketLifecycle `

File: restapi/user_buckets_lifecycle.go [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+	return lifecycleBucketsResponse, nil
+}
+
+// getBucketLifecycle() gets lifecycle lists for a bucket from MinIO API and returns their implementations
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:112]: wrong comments

File: restapi/user_buckets_lifecycle.go [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+	var err2 *probe.Error
+	lfcCfg, err2 = opts.ToConfig(lfcCfg)
+	if err2.ToGoError() != nil {
+		return err2.ToGoError()
+	}
+	if err = client.setBucketLifecycle(ctx, params.BucketName, lfcCfg); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:152]:
```suggestion
	return client.setBucketLifecycle(ctx, params.BucketName, lfcCfg)
```

File: restapi/user_buckets_lifecycle.go [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// getListBucketsResponse performs getBucketLifecycle() and serializes it to the handler's output
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:155]: wrong commenta

File: restapi/user_buckets_lifecycle.go [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+	mClient, err := newMinioClient(session)
+	if err != nil {
+		return prepareError(err)
+	}
+	// create a minioClient interface implementation
+	// defining the client to be used
+	minioClient := minioClient{client: mClient}
+
+	fmt.Println(params.BucketName)
+	fmt.Println(minioClient.listBucketsWithContext(ctx))
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:167]:
```suggestion
```

File: restapi/user_buckets_lifecycle.go [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+	err = addBucketLifecycle(ctx, minioClient, params)
+	if err != nil {
+		fmt.Println("zxczxc")
```
Comment (cesnietor) [inline restapi/user_buckets_lifecycle.go:171]:
```suggestion
```

File: swagger.yml [hunk 24]
```diff
@@ -1586,6 +1651,103 @@ paths:
...
+          schema:
+            $ref: "#/definitions/tier"
+        default:
+          description: Generic error response.
+          schema:
+            $ref: "#/definitions/error"
+      tags:
+        - AdminAPI
+
+  /admin/tiers/{type}/{name}:
```
Comment (cesnietor) [inline swagger.yml:1690]: can the type have a `/`?

File: swagger.yml [hunk 25]
```diff
@@ -3919,6 +4081,95 @@ definitions:
...
+        $ref: "#/definitions/expirationResponse"
+      transition:
+        $ref: "#/definitions/transitionResponse"
+      tags:
+        type: array
+        items:
+          $ref: "#/definitions/lifecycleTag"
+
+
+  addBucketLifecycle:
```
Comment (cesnietor) [inline swagger.yml:4142]: should all properties be non required? you can potentially create a Lifecycle like `{}`

**Review by cesnietor — CHANGES_REQUESTED** [review]
Please add some basic tests

File: go.mod [hunk 26]
```diff
@@ -18,19 +18,25 @@ require (
...
 	gopkg.in/yaml.v2 v2.3.0
 	k8s.io/api v0.20.2
 	k8s.io/apimachinery v0.20.2
 	k8s.io/client-go v0.20.2
 )
+
+replace (
+	github.com/minio/mc v0.0.0-20201220181029-41c804b179de => github.com/krisis/mc v0.0.0-20210201232942-01a91b485075
+	github.com/minio/minio v0.0.0-20210128013121-e79829b5b368 => github.com/minio/minio v0.0.0-20210209004334-fbc470fbb427
+)
```
Comment (cesnietor) [inline go.mod:42]: remove replace

**Review by cesnietor — CHANGES_REQUESTED** [review]

File: swagger.yml (hunk 24 above)
Comment (bexsoft) [reply swagger.yml:1695]: no, it can only be gcs, azure or s3. anything else is not supported

File: swagger.yml (hunk 25 above)
Comment (cesnietor) [reply swagger.yml:4147]: how did this get resolved?

Comment (bexsoft) [reply swagger.yml:4145]: We added a manual validation to this. swagger v2.0 can't handle all the cases that we have for this

**Comment by bexsoft** [comment]
closed in favor of #615

**Pull request closed without merging**